src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -0,0 +1,731 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+
+#include <ripple/nodestore/impl/DatabaseShardImp.h>
+#include <ripple/app/ledger/InboundLedgers.h>
+#include <ripple/app/ledger/Ledger.h>
+#include <ripple/basics/chrono.h>
+#include <ripple/basics/random.h>
+#include <ripple/nodestore/Manager.h>
+#include <ripple/protocol/HashPrefix.h>
+
+namespace ripple {
+namespace NodeStore {
+
+DatabaseShardImp::DatabaseShardImp(Application& app,
+    std::string const& name, Stoppable& parent, Scheduler& scheduler,
+        int readThreads, Section const& config, beast::Journal j)
+    : DatabaseShard(name, parent, scheduler, readThreads, j)
+    , app_(app)
+    , config_(config)
+    , dir_(get<std::string>(config, "path"))
+    , maxDiskSpace_(get<std::uint64_t>(config, "max_size_gb") << 30)
+{
+}
+
+DatabaseShardImp::~DatabaseShardImp()
+{
+    stopThreads();
+}
+
+bool
+DatabaseShardImp::init()
+{
+    using namespace boost::filesystem;
+    {
+        // Validate backend
+        Factory* f;
+        {
+            std::string const type(get<std::string>(config_, "type"));
+            if (type.empty() || !(f = Manager::instance().find(type)))
+            {
+                JLOG(j_.error()) <<
+                    "Invalid shard store type specified";
+                return false;
+            }
+        }
+
+        std::string tmpDir;
+        auto i = 0;
+        do
+            tmpDir = "TMP" + std::to_string(i++);
+        while (is_directory(dir_ / tmpDir));
+
+        auto d = dir_ / tmpDir;
+        auto config {config_};
+        config.set("path", d.string());
+        {
+            fdLimit_ = f->createInstance(NodeObject::keyBytes,
```

Comment:
Is all the temp directory logic a round-about way to get `fdLimit_`?

Reply:
Also to determine if the backend is permanent storage, not too pretty. I will extend the `Factory` class and improve this.